index.bs

```diff
@@ -291,22 +300,34 @@ The <dfn method for="DeviceOrientationEvent">requestPermission()</dfn> method st
 <div algorithm>
 To <dfn>fire an orientation event</dfn> given a <var>event name</var> (a string), <var>window</var> (a {{Window}}) and <var>absolute</var> (a boolean):
 
+  1. Let <var>topLevelTraversable</var> be <var>window</var>'s <a for=Window>navigable</a>'s <a for=navigable>top-level traversable</a>.
+  1. Let <var>platformOrientation</var> be null.
   1. If <var>absolute</var> is false:
-    1. Let <var>orientation</var> be the device's <a>relative orientation</a> in the tridimensional plane.
+      1. If <var>topLevelTraversable</var>'s <a>virtual sensor mapping</a> <a for=map>contains</a> "<code>relative-orientation-euler</code>":
+          1. Let <var>virtualSensor</var> be <var>topLevelTraversable</var>'s <a>virtual sensor mapping</a>["<code>relative-orientation-euler</code>"].
+          1. If <var>virtualSensor</var>'s <a>can provide readings flag</a> is true, then set <var>platformOrientation</var> to the latest readings provided to <var>virtualSensor</var>, so that its "<code>alpha</code>" is the angle along the Z axis, "<code>beta</code>" is the angle around the X' axis, and "<code>gamma</code>" is the angle around the Y'' axis.
```

Comment:
The explicit definition of the axis mapping seems redundant given that we do it again below. This might actually be a good way to avoid needing to specify the Euler angle virtual sensor types at all, since we're specifying that it gets mapped to "intrinsic Tait-Bryan angles Z - X' - Y''" based on whatever the platform gave us anyways.

Reply:
By "specifying the Euler angle virtual sensor types at all" do you mean the bit here after "the latest readings provided to virtualSensor" or are you referring to the virtual sensor type definitions in the Automation section? The latter are required (along with the parsing algorithm) for this spec to integrate properly with the Generic Sensor automation section.

Reply:
I've reordered the steps here, which might have addressed your suggestion as well.

Reply:
The reordering helps, but my thought here was that we could say something along the lines of,

1. Let _orientation_ be null.
1. If the virtual sensor mapping contains _virtualSensorType_,
    1.  If the can provide readings flag is true, let _orientation_ be the latest readings provided to the virtual sensor.
1. Otherwise... (the existing logic which sets orientation based on the device's actual orientation)
1. (the existing logic which takes _orientation_ and represents it as Tait-Bryan angles)

The point here being that since _orientation_ is already somewhat abstract we could say that in the automation case it gets converted from a quaternion and in the non-automation case it gets converted from whatever form the platform provides (which is how it already works).

Reply:
Ack. Just one thing: in the automation case, no quaternions are involved. The idea is to get those 3 doubles (alpha, beta, gamma) with no specific meaning we get via WebDriver and forward them to an absolute/relative-orientation-euler sensor, which eventually provides them to script.

Reply:
If we did this then we could use the same virtual sensor type for the old and new orientation APIs.

Reply:
Sorry, I still don't follow. I don't see how the two APIs would be able to use the same format of sensor readings, and just like these values are just forwarded around until they reach script, so are the quaternion readings for the Orientation Sensor API (i.e. there's no sensor fusion involved in the implementation of either automation API).

Reply:
You can convert between the two formats, so my proposal is that we support one virtual sensor, which we choose to either be fed values as Euler angles or a quaternion it doesn't really matter, and the reading provided to that sensor is fed to both APIs in the same way that creating a virtual accelerometer creates `devicemotion` events and `reading` events on an `Accelerometer` instance.

Reply:
> the same way that creating a virtual accelerometer creates `devicemotion` events and `reading` events on an `Accelerometer` instance.

Right, I think I understand your reasoning and how it'd be useful to be able to control "legacy" `deviceorientation{absolute}` as well as `{Absolute,Relative}OrientationSensor` readings with the same virtual sensor type.

My main concern is that while both `devicemotion` and `Accelerometer` provide readings with the same unit, in this case the conversion between quaternions and the specific set of intrinsic Euler angles that would require this specification to:

1. Take alpha/beta/gamma in "parse orientation data reading", as otherwise the burden of converting Euler angles to quaternions would fall on users.
1. Specify all the math for how these 3 Euler angles are supposed to be converted to a quaternion.
1. Later specify all the math for how to convert the quaternion data back to Euler angles when it is time to provide readings to script.

_(or vice-versa if we choose the canonical format to be Euler angles)_

Going from Euler angles to quaternions is [somewhat trivial](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/sensor/sensor_inspector_agent.cc;l=27;drc=047c7dc4ee1ce908d7fea38ca063fa2f80f92c77), but the other way around requires several steps as seen in Chromium ([here](https://source.chromium.org/chromium/chromium/src/+/main:services/device/generic_sensor/orientation_euler_angles_fusion_algorithm_using_quaternion.cc;l=23;drc=098756533733ea50b2dcb1c40d9a9e18d49febbe) and [here](https://source.chromium.org/chromium/chromium/src/+/main:services/device/generic_sensor/orientation_util.cc;l=16;drc=098756533733ea50b2dcb1c40d9a9e18d49febbe)), [WebKit](https://github.com/WebKit/WebKit/blob/9a3f6dd4c6a6ec871f1fe462628902fe5756cb48/Source/WebCore/platform/ios/WebCoreMotionManager.mm#L227) and [Gecko](https://searchfox.org/mozilla-central/rev/12ea2c521cdd071a6d25b0894f31f8f23b18b76a/dom/system/nsDeviceSensors.cpp#240).

It'a also possible to just be hand-wavy and say "convert" or "represent this quaternion as Euler angles", but I thought the idea was to try to be less hand-wavy here.

Finally, if we do go down this route I guess it doesn't make sense to have "{absolute,relative}-orientation-euler" virtual sensor types, this spec would use/define the "{absolute,relative}-orientation" types instead and have a parsing algorithm that tries to use either Euler angles or quaternions depending on the values provided, right?

Reply:
In the end I'm not really sure which format will be easiest for developers but picking one is better than having two virtual sensors because it provides forward-compatibility for tests should developers adopt the Generic Sensors API. Given the broader adoption of this API it's likely easier if we define the automation in terms of Euler angles so that the values you put into the automation API match what you get in the `deviceorientation` event. That is even though quaternions are "more correct" and most implementations already have the conversion code because of the Android sensor API. It also helps that the formula for converting between Euler angles and a quaternion is the simplest to specify.

Reply:
(https://github.com/w3c/deviceorientation/pull/124/commits/a9c8aa3b8ab8a93198bfc79ba2a62b987821be04 contains the latest commit before I rebased on top of the latest permissions-related changes. The current version of this PR has a single commit on top of the latest main branch)

Here's a possible approach that works with this spec and Orientation Sensor:
- This spec reads Euler angles in the Automation section, and the "fire orientation event" section uses `readings["alpha"]`, `readings["beta"]` and `readings["gamma"]`.
- This spec defines the "{relative,absolute}-orientation" virtual sensor type (no "-euler" at the end) which is shared with the Orientation Sensor spec.
- The Orientation Sensor spec builds on top of this spec's Automation section.
    - It defines algorithms for converting between Euler angles and quaternions.
    - It continues to read and store readings in the "quaternion" entry of the readings map.
    - It modifies the "parse orientation data reading" algorithm so that it:
        - Always sets a "quaternion" key in addition to "alpha"/"beta"/"gamma".
        - If the user-provided reading has a "quaternion" property, it will be preferred over the Euler readings, which will be derived from the quaternion internally.
        - Otherwise, it will derive a quaternion from the Euler angles and set the "quaternion" entry accordingly.

This way, users can specify "{absolute,relative}-orientation" readings by using either quaternions or Euler angles, and all the required conversions are done transparently so that users can use either format if they are using an implementation that supports both.

w3c/orientation-sensor#83 has the Orientation Sensor side of this idea.

Reply:
For posterity: I've just merged w3c/orientation-sensor#83, but the final approach differs from what I described above. Euler angles are converted to quaternions and the "quaternion" key is set, but there's no way to provide quaternions directly and have them converted to Euler angles.

See the discussion in https://github.com/w3c/orientation-sensor/pull/83#discussion_r1421145364